--- a/lib/redmine/codeset_util.rb
+++ b/lib/redmine/codeset_util.rb
@@ -35,9 +35,13 @@
       enc = (encoding.presence || "UTF-8")
       if enc.casecmp("UTF-8") != 0
-        str.force_encoding(enc)
-        str = str.encode("UTF-8", :invalid => :replace,
-              :undef => :replace, :replace => '?')
+        begin
+          str.force_encoding(enc)
+          str = str.encode("UTF-8", :invalid => :replace,
+                :undef => :replace, :replace => '?')
+        rescue Encoding::ConverterNotFoundError, ArgumentError
+          str = replace_invalid_utf8(str)
+        end
       else
         str = replace_invalid_utf8(str)
       end
