Index: app/views/issues/show.html.erb
===================================================================
--- app/views/issues/show.html.erb	(Revision 17840)
+++ app/views/issues/show.html.erb	(Arbeitskopie)
@@ -117,6 +117,13 @@
 <%= render :partial => 'relations' %>
 </div>
 <% end %>
+<% if User.current.allowed_to?(:add_issue_watchers, @project) ||
+    (@issue.watchers.present? && User.current.allowed_to?(:view_issue_watchers, @project)) %>
+  <hr />
+  <div id="watchers">
+    <%= render :partial => 'watchers/watchers', :locals => {:watched => @issue} %>
+  </div>
+<% end %>
 
 </div>
 
@@ -148,12 +155,6 @@
 <% content_for :sidebar do %>
   <%= render :partial => 'issues/sidebar' %>
 
-  <% if User.current.allowed_to?(:add_issue_watchers, @project) ||
-    (@issue.watchers.present? && User.current.allowed_to?(:view_issue_watchers, @project)) %>
-    <div id="watchers">
-      <%= render :partial => 'watchers/watchers', :locals => {:watched => @issue} %>
-    </div>
-  <% end %>
 <% end %>
 
 <% content_for :header_tags do %>
Index: app/views/watchers/_watchers.html.erb
===================================================================
--- app/views/watchers/_watchers.html.erb       2025-05-11 22:35:09.480514580 +0300
+++ app/views/watchers/_watchers.html.erb       2025-05-11 22:34:49.035803000 +0300
@@ -9,9 +9,9 @@
 <% end %>

 <% if User.current.allowed_to?(:"view_#{watched_klass_name}_watchers", watched.project) %>
-<h3><%= l(:"label_#{watched_klass_name}_watchers") %> (<%= watched.watcher_users.size %>)</h3>
+<p><strong><%= l(:"label_#{watched_klass_name}_watchers") %> (<%= watched.watcher_users.size %>)</strong></p>

 <%= watchers_list(watched) %>
 <% else %>
-<h3><%= l(:"label_#{watched_klass_name}_watchers") %></h3>
+<p><strong><%= l(:"label_#{watched_klass_name}_watchers") %></strong></p>
 <% end %>
 